### PR TITLE
Allow noninteractive icon

### PR DESCRIPTION
**Pull request by bdantas**
Jon,

Most of the time it's great to be able to control the icon with a named pipe (e.g., to monitor wifi strength or battery charge). However, sometimes I just want to create a simple, non-interactive icon to show that something is on (e.g., VPN). For these situations, it's easier to just specify the tooltip right away (one step) without having to create a pipe, call mktrayicon, then specify the tooltip via the pipe (three steps). When such an icon is no longer needed, that instance of mktrayicon can simply be killed using `pkill -f 'mktrayicon.*<icon_name>'`. There is also less overhead when the icon is no longer needed, since there is no obsolete pipe to delete.

Hopefully you agree that this is a nice feature. Here's the patch.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,9 +3,20 @@
 `mktrayicon` is a simple proxy program that lets you create and modify system
 tray icons without having to deal with a graphical toolkit like GTK.
 
-`mktrayicon` expects to be given a name pipe (FIFO) file path when it is
-started, and you control your status icon by writing to this named pipe. *Note
-that the pipe should already be created before you call `mktrayicon`*.
+`mktrayicon` can be used two ways: To create an icon that is controlled by a named pipe
+or, more simply, to create a non-interactive icon using the specified tooltip string (which can be an empty string if you don't need a tooltip).
```
Comment (jonhoo) [inline README.md]: Could you wrap this line and the ones above to match the existing text wrapping?

Comment (jonhoo) [inline README.md]: Does it explicitly have to be the empty string (i.e., `-t ""`), or can it be simply omitting the `-t`?

File: README.md [hunk 2]
```diff
@@ -3,9 +3,20 @@
...
+or, more simply, to create a non-interactive icon using the specified tooltip string (which can be an empty string if you don't need a tooltip). 
+
+In all, there are three ways of calling mktrayicon:
+```
+$ mktrayicon <FIFO> &
+  or
+$ mktrayicon -i <ICON> <FIFO> &
+  or
+$ mktrayicon -i <ICON> -t <TOOLTIP> &
+```
```
Comment (jonhoo) [inline README.md]: Is
```console
$ mktrayicon -i <ICON> &
```
supported too?

File: mktrayicon.c [hunk 3]
```diff
@@ -216,26 +212,34 @@ int main(int argc, char **argv)
 	char *start_icon = "none";
 	FILE *fifo;
 	GThread *reader;
-
```
Comment (jonhoo) [inline mktrayicon.c]: Why this change?

File: mktrayicon.c [hunk 4]
```diff
@@ -216,26 +212,34 @@ int main(int argc, char **argv)
...
 	if (argc == 1) {
-		printf("Usage: %s [-i ICON] FIFO\n", *argv);
-		printf("Listen to FIFO for system tray icon specifications\n");
+		printf("Usage: %s [-i ICON] FIFO --or-- %s -i ICON -t TOOLTIP\n", *argv, *argv);
+		printf("Create a system tray icon as specified\n");
 		printf("\n");
 		printf("  -i ICON\tUse the specified ICON when initializing\n");
+		printf("  -t TOOLTIP\tUse the specified TOOLTIP when initializing\n");
 		printf("\n");
 		printf("Report bugs at https://github.com/jonhoo/mktrayicon\n");
```
Comment (jonhoo) [inline mktrayicon.c]: Maybe add a line here along the lines of
> If `FIFO` is not specified, the tray icon will instead persist for as long as `mktrayicon` runs.

File: README.md [hunk 5]
```diff
@@ -3,9 +3,20 @@
 `mktrayicon` is a simple proxy program that lets you create and modify system
 tray icons without having to deal with a graphical toolkit like GTK.
 
-`mktrayicon` expects to be given a name pipe (FIFO) file path when it is
-started, and you control your status icon by writing to this named pipe. *Note
-that the pipe should already be created before you call `mktrayicon`*.
+`mktrayicon` can be used two ways: To create an icon that is controlled by a named pipe
```
Comment (jonhoo) [inline README.md]: I wonder if an easier way to phrase this is "if a FIFO is not provided, `mktrayicon` will run until killed".

File: mktrayicon.c [hunk 6]
```diff
@@ -216,26 +212,34 @@ int main(int argc, char **argv)
...
 		return 0;
 	}
 
+	if (strcmp(argv[1], "-i") == 0) {
+		start_icon = argv[2];
```
Comment (jonhoo) [inline mktrayicon.c]: Come to think of it, we should probably emit an error if `argv[2]` isn't set here.

File: mktrayicon.c [hunk 7]
```diff
@@ -216,26 +212,34 @@ int main(int argc, char **argv)
...
+	}
+
 	icon = create_tray_icon(start_icon);
-	reader = g_thread_new("watch_fifo", watch_fifo, argv[argc-1]);
+
+	if (strcmp(argv[3], "-t") == 0) { /* icon is non-interactive, use given tooltip */
+		gtk_status_icon_set_tooltip_text(icon, argv[4]);
+	}
+	else { /* icon is interactive, listen to the specified named pipe */
```
Comment (jonhoo) [inline mktrayicon.c]: I think you want a branch here along the lines of
```
else if (argc == 3) {
```
to capture the case where the user provides `-i icon` but no `FIFO` and no `-t`.

File: mktrayicon.c [hunk 8]
```diff
@@ -216,26 +212,34 @@ int main(int argc, char **argv)
...
+		reader = g_thread_new("watch_fifo", watch_fifo, argv[argc-1]);
```
Comment (jonhoo) [inline mktrayicon.c]: This can probably be `argv[3]` now.

File: mktrayicon.c [hunk 9]
```diff
@@ -216,26 +212,34 @@ int main(int argc, char **argv)
...
 
+	if (strcmp(argv[1], "-i") == 0) {
+		start_icon = argv[2];
+	}
+
 	icon = create_tray_icon(start_icon);
-	reader = g_thread_new("watch_fifo", watch_fifo, argv[argc-1]);
+
+	if (strcmp(argv[3], "-t") == 0) { /* icon is non-interactive, use given tooltip */
+		gtk_status_icon_set_tooltip_text(icon, argv[4]);
```
Comment (jonhoo) [inline mktrayicon.c]: We should error if `argv[4]` isn't set.

**Comment by jonhoo** [comment]
Seems like a great idea!

File: README.md (hunk 1 above)
Comment (bdantas) [reply README.md]: -t is what signals that the icon is non-interactive. Both `mktrayicon -i <icon> -t ""` and `mktrayicon -i <icon> -t` work.

File: README.md (hunk 2 above)
Comment (bdantas) [reply README.md]: No, but we can do that. I just didn't want to change things around too much.

File: mktrayicon.c (hunk 3 above)
Comment (bdantas) [reply mktrayicon.c]: That was inadvertent.

File: README.md (hunk 1 above)
Comment (bdantas) [reply README.md]: Ok. Everything is now wrapped at 80 columns.

File: mktrayicon.c (hunk 9 above)
Comment (bdantas) [reply mktrayicon.c]: If argv[3] is -t not followed by a string, it still works. Less typing to leave blank than to type `""`

**Comment by bdantas** [comment]
Let me chew on this for an hour or so. I'll let you know when things look better.

**Comment by bdantas** [comment]
I've done as much as I can for today, I think.

Everything works except `mktrayicon -i <ICON> &` which segfaults for some reason.

File: mktrayicon.c (hunk 4 above)
Comment (jonhoo) [reply mktrayicon.c]: That line should probably not be indented. Maybe place it after the `\n`, and capitalize the first letter?

File: mktrayicon.c [hunk 10]
```diff
@@ -237,9 +238,13 @@ int main(int argc, char **argv)
 		gtk_status_icon_set_tooltip_text(icon, argv[4]);
 	}
 
-	if (/* test to see if last arg is a pipe */) {
+	/* test if last argument is a pipe */
+	char *command = malloc(1024*sizeof(char));
+	sprintf(command, "test -p %s", argv[argc-1]);
```
Comment (jonhoo) [inline mktrayicon.c]: I don't think you need this? If the last argument is a `-t`, or the last argument followed a `-i` or `-t`, then we know it isn't a pipe.

Comment (bdantas) [reply mktrayicon.c]: It's an eyesore, I know. But I can't think of anything else that would be foolproof. We could test if next-to-last argument contained a `-` but that would fail if call was `mktrayicon <FIFO>`

Comment (jonhoo) [reply mktrayicon.c]: Here's what I would do:
 - Set `args = 1`
 - If `-i`, do `args += 2`
 - If `-t`, do `args += 2`
 - If `args < argc`, `argv[argc - 1]` is a FIFO

Comment (bdantas) [reply mktrayicon.c]: Cool. Done. 

For some reason it still segfaults on `mktrayicon -i <ICON> &`. Any ideas?

Comment (bdantas) [reply mktrayicon.c]: Oh, I think argv[3] becomes a problem if there are only two arguments.

**Comment by bdantas** [comment]
Ok, I think it's ready.

File: mktrayicon.c (hunk 4 above)
Comment (jonhoo) [reply mktrayicon.c]: I'd still like this to be addressed :)

File: mktrayicon.c [hunk 11]
```diff
@@ -220,22 +216,36 @@ int main(int argc, char **argv)
...
+		printf("  if a FIFO is not provided, mktrayicon will run until killed\n");
 		printf("\n");
 		printf("Report bugs at https://github.com/jonhoo/mktrayicon\n");
 		return 0;
 	}
 
+	int args = 1;
+
+	if (strcmp(argv[1], "-i") == 0) {
+		start_icon = argv[2];
```
Comment (jonhoo) [inline mktrayicon.c]: Will this also segfault if I run `mktrayicon -i` ? I think we should explicitly warn if `argc < 3` here.

File: mktrayicon.c [hunk 12]
```diff
@@ -220,22 +216,36 @@ int main(int argc, char **argv)
...
+		args += 2;
+	}
+
+	if (argc > 3 && strcmp(argv[3], "-t") == 0) {
```
Comment (jonhoo) [inline mktrayicon.c]: This won't quite work, as it breaks the logic needed for `args < argc` below. I think it should be:
```c
if (strcmp(argv[3], "-t") == 0) {
    if (argc >= 5) {
        gtk_status...
    }
    args += 2;
}
```

File: mktrayicon.c [hunk 13]
```diff
@@ -220,22 +216,36 @@ int main(int argc, char **argv)
...
 	gtk_init(&argc, &argv);
 
-	if (argc == 4 && strcmp(argv[1], "-i") == 0) {
-		start_icon = argv[2];
-	}
-
 	if (argc == 1) {
-		printf("Usage: %s [-i ICON] FIFO\n", *argv);
-		printf("Listen to FIFO for system tray icon specifications\n");
+		printf("Usage: %s [-i ICON] [-t TOOLTIP] [FIFO]\n", *argv);
```
Comment (jonhoo) [inline mktrayicon.c]: This is now a little weird, because we actually _require_ `-i` if you provide `-t`. Maybe we should add a real argument parsing lib :p

Comment (bdantas) [reply mktrayicon.c]: True. 
I don't have time for it today, will get around to it when I can. I'll be going away on vacation with family this weekend, so it probably won't be until late next week.

File: mktrayicon.c [hunk 14]
```diff
@@ -210,14 +211,16 @@ static GtkStatusIcon *create_tray_icon(char *start_icon)
...
+	char *pipe = NULL;
 	FILE *fifo;
 	GThread *reader;
 
 	XInitThreads(); /* see http://stackoverflow.com/a/18690540/472927 */
 	gtk_init(&argc, &argv);
 
 	if (argc == 1) {
-		printf("Usage: %s [-i ICON] [-t TOOLTIP] [FIFO]\n", *argv);
+		printf("Usage: %s [-i ICON] [-t TOOLTIP] [-p FIFO]\n", *argv);
```
Comment (jonhoo) [inline mktrayicon.c]: I'd prefer to keep `FIFO` as a positional argument. I believe `getopt` still supports that.

File: README.md [hunk 15]
```diff
@@ -6,6 +6,21 @@ tray icons without having to deal with a graphical toolkit like GTK.
 `mktrayicon` can be used two ways: To create an icon that is controlled by a 
 named pipe or, more simply, to create a non-interactive icon.
 
+Calls to mktrayicon are flexible. All of these work:
```
Comment (jonhoo) [inline README.md]: On second thought, I feel like this list probably isn't useful. Instead, I think it's sufficient to say that "if you do not pass in a pipe, `mktrayicon` will run with the given icon and tooltip text until terminated".

**Comment by bdantas** [comment]
Alright, Jon. I'm happy with this. Sure we can add sanity checks on the options, but I don't want to overcomplicate things.

Feel free to merge it and add any final touches.

**Comment by bdantas** [comment]
Jon, please take a peek at it. All good?

File: mktrayicon.c [hunk 16]
```diff
@@ -214,28 +211,57 @@ static GtkStatusIcon *create_tray_icon(char *start_icon)
...
 	if (argc == 1) {
-		printf("Usage: %s [-i ICON] FIFO\n", *argv);
-		printf("Listen to FIFO for system tray icon specifications\n");
+		printf("Usage: %s [-i ICON] [-t TOOLTIP] [FIFO]\n", *argv);
+		printf("Create a system tray icon as specified\n");
 		printf("\n");
 		printf("  -i ICON\tUse the specified ICON when initializing\n");
+		printf("  -t TOOLTIP\tUse the specified TOOLTIP when initializing\n");
+		printf("  if a FIFO is provided, it should be the last argument\n");
+		printf("  if a FIFO is not provided, mktrayicon will run until killed\n");
```
Comment (jonhoo) [inline mktrayicon.c]: I think I'd prefer for these lines to not be inline with the argument list, but rather after the `printf("\n")` below, without the indentation, with a capital first letter.

File: mktrayicon.c [hunk 17]
```diff
@@ -214,28 +211,57 @@ static GtkStatusIcon *create_tray_icon(char *start_icon)
...
+
+
 	icon = create_tray_icon(start_icon);
-	reader = g_thread_new("watch_fifo", watch_fifo, argv[argc-1]);
+
+	if (tooltip) {
+		gtk_status_icon_set_tooltip_text(icon, tooltip);
+	}
+
+	if (parsed_args < argc) {
```
Comment (jonhoo) [inline mktrayicon.c]: I'm surprised `getopt` doesn't have support for positional arguments?

Comment (bdantas) [reply mktrayicon.c]: It does, but is less than intuitive:
https://www.gnu.org/software/libc/manual/html_node/Example-of-Getopt.html#Example-of-Getopt

I edited the code to use this method. Now the pipe does not have to be the last argument.

File: mktrayicon.c (hunk 16 above)
Comment (bdantas) [reply mktrayicon.c]: Ok.

**Comment by bdantas** [comment]
It turns out that the value of `optind` after the while loop is equal to the number of parsed arguments. So we can use it exactly as we would use the `parsed_args` variable I had in there before.

**Comment by bdantas** [comment]
I understand it now: getopt moves positional arguments (if there are any) to the end of the argv array and does not count them.

If there are only option args, optind will be equal to argc after getopt's switch.

If there are positional arguments, argv[optindex] is the first, argv[optindex+1] is the second, etc.

Bottomline is that you don't need a flag for FIFO and it can be anywhere in the command as long as it is the first positional parameter (i.e., the first argument not preceded by a flag) :)

**Comment by jonhoo** [comment]
Looks great, thank you!

**Comment by bdantas** [comment]
You're welcome. Thanks again for mktrayicon, and for insisting on clean code. It pushed me to learn something new. Happy hacking!
